lib/nexthop.c

```diff
@@ -36,40 +36,124 @@
 DEFINE_MTYPE_STATIC(LIB, NEXTHOP, "Nexthop")
 DEFINE_MTYPE_STATIC(LIB, NH_LABEL, "Nexthop label")
 
-/* check if nexthops are same, non-recursive */
-int nexthop_same_no_recurse(const struct nexthop *next1,
-			    const struct nexthop *next2)
+static int nexthop_labels_cmp(const struct nexthop *nh1,
+			      const struct nexthop *nh2)
 {
-	if (next1->type != next2->type)
+	const struct mpls_label_stack *nhl1 = NULL;
+	const struct mpls_label_stack *nhl2 = NULL;
+
+	nhl1 = nh1->nh_label;
+	nhl2 = nh2->nh_label;
+
+	/* No labels is a match */
+	if (!nhl1 && !nhl2)
 		return 0;
 
-	switch (next1->type) {
+	if (nhl1 && !nhl2)
+		return 1;
+
+	if (nhl2 && !nhl1)
+		return -1;
+
+	if (nhl1->num_labels > nhl2->num_labels)
+		return 1;
+
+	if (nhl1->num_labels < nhl2->num_labels)
+		return -1;
+
+	return memcmp(nhl1->label, nhl2->label, nhl1->num_labels);
+}
+
+static int nexthop_g_addr_cmp(enum nexthop_types_t type,
+			      const union g_addr *addr1,
+			      const union g_addr *addr2)
+{
+	int ret = 0;
+
+	switch (type) {
 	case NEXTHOP_TYPE_IPV4:
 	case NEXTHOP_TYPE_IPV4_IFINDEX:
-		if (!IPV4_ADDR_SAME(&next1->gate.ipv4, &next2->gate.ipv4))
-			return 0;
-		if (next1->ifindex && (next1->ifindex != next2->ifindex))
-			return 0;
+		ret = IPV4_ADDR_CMP(&addr1->ipv4, &addr2->ipv4);
+		break;
+	case NEXTHOP_TYPE_IPV6:
+	case NEXTHOP_TYPE_IPV6_IFINDEX:
+		ret = IPV6_ADDR_CMP(&addr1->ipv6, &addr2->ipv6);
 		break;
 	case NEXTHOP_TYPE_IFINDEX:
-		if (next1->ifindex != next2->ifindex)
-			return 0;
+	case NEXTHOP_TYPE_BLACKHOLE:
+		/* No addr here */
 		break;
+	}
+
+	return ret;
+}
+
+static int nexthop_gateway_cmp(const struct nexthop *nh1,
+			       const struct nexthop *nh2)
+{
+	return nexthop_g_addr_cmp(nh1->type, &nh1->gate, &nh2->gate);
+}
+
+static int nexthop_source_cmp(const struct nexthop *nh1,
+			      const struct nexthop *nh2)
+{
+	return nexthop_g_addr_cmp(nh1->type, &nh1->src, &nh2->src);
+}
+
+int nexthop_cmp(const struct nexthop *next1, const struct nexthop *next2,
+		bool ignore_labels)
```

Comment:
oh man - you're going to hate this: I dislike adding this to the signature - especially if "false" means "yes, do everything." I think I understand Renato's example - but I think the issue is that we don't properly construct a complete nexthop down in the netlink code in the 'delete' path - we don't even try to look for labels there. in the KERNEL delete path, we might do better to just consider the uniqueness to be type + metric, and skip the nexthop comparison.

If there really are cases where two nexthops that don't have the same label stack are considered "the same", I guess it'd be ok to have a nexthop_cmp_no_labels() or nexthop_same_no_labels() api for that. but it's hard to understand how that could really be correct - aren't those really not the same nexthop?

Reply:
Actually, in reference to @rwestphal 's example.

It looks like in `zebra/rt_netlink.c` we are not getting the labels from the kernel with a route delete. We just ignore them. So maybe that would fix the issue he is talking about it without adding this code edge case.

Reply:
Deleted my comment once I saw yours. Could I not just add code in the kernel delete path to actually parse the labels and add them to it?

Reply:
If I do that, I believe I can just get rid of the edge case bool all together.

Reply:
> [...] in the KERNEL delete path, we might do better to just consider the uniqueness to be type + metric, and skip the nexthop comparison.

Unfortunately I think this wouldn't be possible. The kernel doesn't use route replace semantics when informing the userspace about the removal of a nexthop of an ECMP route. This means that, upon receipt of a RTM_DELETE message, we need to find the exact nexthop we need to remove from the RIB.

> Could I not just add code in the kernel delete path to actually parse the labels and add them to it?
> Deleted my comment once I saw yours. Could I not just add code in the kernel delete path to actually parse the labels and add them to it?

That would solve the problem, but you'd need to do the same thing in _zebra/kernel_socket.c_. The problem, however, is that we don't know if other kernels (e.g. OpenBSD) send route delete messages containing MPLS labels, and this renders this solution a bit fragile.

Also, being able to do partial nexthop comparisons might be useful for other things in the future. One example that comes to mind is implementing a "nexthop" option to the "show ip route" command. Some vendors implement this option, which allows you to see only the routes that point to a specific nexthop, regardless if it has MPLS labels or not.

Reply:
I didn't mean to imply that I didn't want to offer a more limited comparison - just that:
a) the specific example had some other aspects to it, and
b) I'd rather offer the special case as a special api, like `nexthop_same_no_labels` so that it's really clear what's happening, and so that every other user of `nexthop_same` doesn't have to figure out how to use an extra boolean arg.

Reply:
Understood. The only problem I see with `nexthop_same_no_labels` is that it can lead to a bit of code duplication, but this might be acceptable in this case.

Other option would be to rename the `ignore_labels` parameter to `flags`, and add flags like `F_NHCMP_IGN_LABELS`. This would give us more flexibility to perform other kinds of comparisons in the future (e.g. ignore ifindexes?), but admittedly we might not need all this flexibility.

I'll approve this PR now as it looks good to me already (the `rib_delete()` bug was fixed), regardless if you want to do further changes to improve the API.

Reply:
Thanks, Renato - all of that sounds sensible.
totally agree about duplication: I'd suggest taking the body of the implementation and making it `_internal()` and `static`. then the public apis - `nexthop_same()` and `nexthop_same_no_labels()` - would just pass through to the shared implementation. If in the future we want more flavors, we can support that too, while having a clear public api that's easy to use correctly.

Reply:
> I'd suggest taking the body of the implementation and making it `_internal()` and `static`. then the public apis - `nexthop_same()` and `nexthop_same_no_labels()` - would just pass through to the shared implementation.

Sounds good to me.